### PR TITLE
Use annotations for operation descriptions

### PR DESCRIPTION
**Pull request by SamCarlberg**
Operations are created via the Guice `Injector`, making it much easier to use dependency injection in operations. Operations are also discovered at startup as long as they have an `@Description` annotation on the class

### CODE REVIEW

**Comment by SamCarlberg** [comment]
As an FYI the branch for CUDA support will be based on this for dependency injection

File: core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java [hunk 1]
```diff
@@ -19,19 +28,45 @@
...
+  @AfterClass
+  public static void tearDownClass() {
+    System.out.println("tearDownClass()");
+    testModule.tearDown();
+  }
+
   @Parameters(name = "{index}: Operation({0})")
   public static Collection<Object[]> data() {
-    EventBus eventBus = new EventBus();
+    System.out.println("data()");
```
Comment (AustinShalit) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: Why did PMD not find this...

Comment (SamCarlberg) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: I don't think PMD runs on test files

File: core/src/main/java/edu/wpi/grip/core/OperationDescription.java [hunk 2]
```diff
@@ -28,6 +28,33 @@
   private final ImmutableSet<String> aliases;
 
   /**
+   * Creates an operation description from a {@link Description @Description} annotation on
+   * an operation subclass.
+   */
+  public static OperationDescription from(Description description) {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/OperationDescription.java]: No can do. Annotation interfaces can't have methods.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
 import edu.wpi.grip.core.Operation;
 import edu.wpi.grip.core.OperationDescription;
 import edu.wpi.grip.core.sockets.InputSocket;
 import edu.wpi.grip.core.sockets.OutputSocket;
 import edu.wpi.grip.core.sockets.SocketHint;
 import edu.wpi.grip.core.sockets.SocketHints;
-import edu.wpi.grip.core.util.Icon;
 
 import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java]: Can you use `javax.inject.Inject` instead?

File: core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java [hunk 4]
```diff
@@ -11,6 +11,7 @@
  * {@link Size}s are converted into this.
  */
 @Immutable
+@PublishableProxy({Point.class, Size.class})
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java]: Do you want to use the constructor parameter to get the type?
Any reason why the type should be specified in the parameter and in the annotation?

File: core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java [hunk 5]
```diff
@@ -19,19 +28,45 @@
...
+
+  static {
+    // Only because @Parameters runs before @BeforeClass... for some weird reason
+    setUpClass();
+  }
+
+  // @Parameters is called before @BeforeClass, so we call it manually in a static initializer block
+  //  @BeforeClass
+  public static void setUpClass() {
+    System.out.println("setUpClass");
```
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: Remove this please.

File: core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java [hunk 6]
```diff
@@ -19,19 +28,45 @@
...
+    testModule = new GripCoreTestModule();
+    testModule.setUp();
+    injector = Guice.createInjector(
+        Modules.override(new GripCoreModule(), new GripNetworkModule(), new GripFileModule())
+            .with(new MockGripNetworkModule(), testModule));
+  }
+
+  @AfterClass
+  public static void tearDownClass() {
+    System.out.println("tearDownClass()");
```
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: Remove this please.

File: core/src/test/java/edu/wpi/grip/core/serialization/CompatibilityTest.java [hunk 7]
```diff
@@ -47,15 +50,17 @@ public void setUp() throws Exception {
...
 
     final EventBus eventBus = injector.getInstance(EventBus.class);
     pipeline = injector.getInstance(Pipeline.class);
     final Project project = injector.getInstance(Project.class);
 
     //Add the operations so that GRIP will recognize them
-    OperationsFactory.create(eventBus).addOperations();
-    //CVOperations.addOperations(eventBus);
+    OperationsFactory.create(eventBus, injector).addOperations();
+    OperationsFactory.createCV(eventBus).addOperations();
```
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/serialization/CompatibilityTest.java]: Why was this commented out before?

File: core/src/test/resources/edu/wpi/grip/projects/testALL.grip [hunk 8]
```diff
@@ -405,9 +405,6 @@
       <grip:Input step="34" socket="1">
         <value>COLOR_BGR2GRAY</value>
       </grip:Input>
-      <grip:Input step="34" socket="2">
-        <value>0.0</value>
-      </grip:Input>
```
Comment (JLLeitschuh) [inline core/src/test/resources/edu/wpi/grip/projects/testALL.grip]: Why was this removed?

File: core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java (hunk 3 above)
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java]: Why?

File: core/src/test/resources/edu/wpi/grip/projects/testALL.grip (hunk 8 above)
Comment (SamCarlberg) [inline core/src/test/resources/edu/wpi/grip/projects/testALL.grip]: It was incompatible with the new version of the operation. I think the compatibility test was disabled for so long that it was never picked up

File: core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java (hunk 4 above)
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java]: We _could_ iterate over all public constructors, but I like how this specifies exactly what's being proxied. It could also run into issues with publicly accessible helper constructors eg `Vector2D(double, double)`

File: core/src/main/java/edu/wpi/grip/core/OperationDescription.java (hunk 2 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/OperationDescription.java]: Okay, thats fine.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java (hunk 3 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java]: > Prefer JSR-330's annotations and Provider interface.

https://github.com/google/guice/wiki/JSR330

File: core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java (hunk 4 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/network/Vector2D.java]: Fine. This makes sense.

File: core/src/test/resources/edu/wpi/grip/projects/testALL.grip (hunk 8 above)
Comment (AustinShalit) [inline core/src/test/resources/edu/wpi/grip/projects/testALL.grip]: 😱

Comment (JLLeitschuh) [inline core/src/test/resources/edu/wpi/grip/projects/testALL.grip]: So somewhere along the way we had a breaking change that we never did a major version bump for?

**Comment by SamCarlberg** [comment]
@JLLeitschuh want to revisit this?

File: core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java (hunk 3 above)
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/FilterLinesOperation.java]: Oh. That's going to be a lot of code changed because both are used everywhere. Should probably add a checkstyle rule for this in a different PR

File: core/src/main/java/edu/wpi/grip/core/Description.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+  /**
+   * All known aliases of the operation. If the name of the operation changes, the previous name
+   * should be here. Defaults to an empty array.
+   */
+  String[] aliases() default {};
+
+  /**
+   * The name of the icon to use to display the operation. If empty ({@code ""}), no icon will be
+   * shown. The icon should be located in {@code /edu/wpi/grip/ui/icons/}.
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Description.java]: Is this the right use of this javadoc annotation? Does it render correctly?

File: core/src/main/java/edu/wpi/grip/core/operations/network/ros/JavaToMessageConverter.java [hunk 10]
```diff
@@ -125,7 +136,7 @@ public void convert(J javaType, Message message, MessageFactory messageFactory)
   }
 
   private abstract static class SimpleConverter<J, M extends Message> extends
-      JavaToMessageConverter<J, M> {
+                                                                      JavaToMessageConverter<J, M> {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/network/ros/JavaToMessageConverter.java]: Why is this whitespace diff here?

File: core/src/main/java/edu/wpi/grip/core/serialization/StepConverter.java [hunk 11]
```diff
@@ -61,7 +62,12 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
         .getOperationByName(operationName);
 
     if (!operationMetaData.isPresent()) {
-      throw new ConversionException("Unknown operation: " + operationName);
+      throw new ConversionException(String.format("Unknown operation: %s. Known operations: %s",
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/serialization/StepConverter.java]: Perhaps `"Unknown operation: %s. did not match any known operations: %s"`

File: core/src/test/java/edu/wpi/grip/core/CoreSanityTest.java [hunk 12]
```diff
@@ -27,7 +27,8 @@ public CoreSanityTest() {
         ManualPipelineRunner.class,
         SubtractionOperation.class,
         Main.class,
-        CoreCommandLineHelper.class
+        CoreCommandLineHelper.class,
+        OperationDescription.class
```
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/CoreSanityTest.java]: Why are you adding this class? Why not simply make it pass?

File: core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java (hunk 5 above)
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: This still needs to be addressed.

File: core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java (hunk 6 above)
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/operations/OperationsTest.java]: This still needs to be addressed.

File: core/src/test/resources/edu/wpi/grip/projects/testALL.grip (hunk 8 above)
Comment (JLLeitschuh) [inline core/src/test/resources/edu/wpi/grip/projects/testALL.grip]: Any update on this?

File: core/src/main/java/edu/wpi/grip/core/operations/Operations.java [hunk 13]
```diff
@@ -1,202 +1,268 @@
...
-             @Named("ntManager") MapNetworkPublisherFactory ntPublisherFactory,
-             @Named("httpManager") MapNetworkPublisherFactory httpPublishFactory,
-             @Named("rosManager") ROSNetworkPublisherFactory rosPublishFactory,
+             @Named("ntManager") MapNetworkPublisherFactory ntManager,
+             @Named("httpManager") MapNetworkPublisherFactory httpManager,
+             @Named("rosManager") ROSNetworkPublisherFactory rosManager,
+             Injector injector,
              FileManager fileManager,
              InputSocket.Factory isf,
              OutputSocket.Factory osf) {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/Operations.java]: Does the `OutputSocket.Factory` still get used anymore?

File: core/src/main/java/edu/wpi/grip/core/Description.java (hunk 9 above)
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/Description.java]: Yeah, it's just a shortcut for `<code>...</code>`

**Comment by JLLeitschuh** [comment]
I cleared up my own review comments and I'll merge this if the CI passes.
